File: wini source/exepciones.py
# exepciones.py
"""Excepciones del intérprete wini"""

class RetornoException(Exception):
    """Excepción para manejar retornos dentro de funciones."""
    def __init__(self, valor):
        self.valor = valor
        super().__init__()

class RomperException(Exception):
    """Excepción para romper bucles."""
    pass

class ContinuarException(Exception):
    """Excepción para continuar en bucles."""
    pass

# === NUEVAS EXCEPCIONES PARA TRY-CATCH ===
class LanzarException(Exception):
    """Excepción para lanzar errores manualmente (lanzar)"""
    def __init__(self, valor, linea=None):
        self.valor = valor
        self.linea = linea
        super().__init__(str(valor))

class ErrorConLinea(Exception):
    """Base para excepciones que tienen línea y columna."""
    def __init__(self, mensaje, linea=None, columna=None, texto=None):
        self.mensaje = mensaje
        self.linea = linea
        self.columna = columna
        self.texto = texto
        super().__init__(mensaje)

    def __str__(self):
        base = self.mensaje
        if self.linea is not None:
            base += f" (línea {self.linea}"
            if self.columna is not None:
                base += f", columna {self.columna}"
            base += ")"
        if self.texto:
            base += f"\n  {self.texto}"
        return base

class SintaxisError(ErrorConLinea):
    """Error de sintaxis con ubicación."""
    pass

class RuntimeError(ErrorConLinea):
    """Error en tiempo de ejecución con ubicación."""
    pass

# Nuevos tipos de error para try-catch
class ErrorTipo(ErrorConLinea):
    """Error de tipo (TypeError)"""
    pass

class ErrorValor(ErrorConLinea):
    """Error de valor (ValueError)"""
    pass

class ErrorIndice(ErrorConLinea):
    """Error de índice (IndexError)"""
    pass

class ErrorAtributo(ErrorConLinea):
    """Error de atributo (AttributeError)"""
    pass

class ErrorImportacion(ErrorConLinea):
    """Error de importación (ImportError)"""
    pass

class ErrorMatematico(ErrorConLinea):
    """Error matemático (ZeroDivisionError, etc)"""
    pass

File: gowini/wini/exepciones.go
package wini

import "fmt"

// ErrorConLinea es la base para errores que tienen línea y columna
type ErrorConLinea struct {
	Mensaje string
	Linea   *int
	Columna *int
	Texto   string
}

func (e *ErrorConLinea) Error() string {
	base := e.Mensaje
	if e.Linea != nil {
		base += fmt.Sprintf(" (línea %d", *e.Linea)
		if e.Columna != nil {
			base += fmt.Sprintf(", columna %d", *e.Columna)
		}
		base += ")"
	}
	if e.Texto != "" {
		base += fmt.Sprintf("\n  %s", e.Texto)
	}
	return base
}

// SintaxisError es un error de sintaxis con ubicación
type SintaxisError struct {
	ErrorConLinea
}

// RuntimeError es un error en tiempo de ejecución con ubicación
type RuntimeError struct {
	ErrorConLinea
}

// ErrorTipo es un error de tipo (TypeError)
type ErrorTipo struct {
	ErrorConLinea
}

// ErrorValor es un error de valor (ValueError)
type ErrorValor struct {
	ErrorConLinea
}

// ErrorIndice es un error de índice (IndexError)
type ErrorIndice struct {
	ErrorConLinea
}

// ErrorAtributo es un error de atributo (AttributeError)
type ErrorAtributo struct {
	ErrorConLinea
}

// ErrorImportacion es un error de importación (ImportError)
type ErrorImportacion struct {
	ErrorConLinea
}

// ErrorMatematico es un error matemático (ZeroDivisionError, etc)
type ErrorMatematico struct {
	ErrorConLinea
}

// RetornoException es una excepción para manejar retornos dentro de funciones
type RetornoException struct {
	Valor interface{}
}

func (e *RetornoException) Error() string {
	return fmt.Sprintf("return %v", e.Valor)
}

// RomperException es una excepción para romper bucles
type RomperException struct{}

func (e *RomperException) Error() string {
	return "break"
}

// ContinuarException es una excepción para continuar en bucles
type ContinuarException struct{}

func (e *ContinuarException) Error() string {
	return "continue"
}

// LanzarException es una excepción para lanzar errores manualmente
type LanzarException struct {
	Valor interface{}
	Linea *int
}

func (e *LanzarException) Error() string {
	if e.Linea != nil {
		return fmt.Sprintf("%v (línea %d)", e.Valor, *e.Linea)
	}
	return fmt.Sprintf("%v", e.Valor)
}
